File: app/src/main/java/com/example/myapplicationbird/GameThread.java
package com.example.myapplicationbird;

import android.graphics.Canvas;
import android.os.SystemClock;
import android.view.SurfaceHolder;

import androidx.core.content.UnusedAppRestrictionsConstants;

public class GameThread extends Thread{

    SurfaceHolder surfaceHolder; //Surfaceholder object reference
    boolean isRunning;
    long startTime,loopTime;
    long DELAY = 33;
    public GameThread(SurfaceHolder){
        this.surfaceHolder = surfaceHolder;
        isRunning = true;
    }
    @Override
    public void run(){
        //looping until the boolean is false
        while(isRunning){
            startTime = SystemClock.uptimeMillis();
            Canvas canvas = surfaceHolder.lockCanvas(dirty:null);
            if(canvas!=null){
                synchronized (surfaceHolder){
                    AppConstants.getGameEndine().updateAndDrawBackgroundImage(canvas);
                    //unlocking the canvas
                    surfaceHolder.unlockCanvasAndPost(canvas);

                }
            }
            //loop time
            loopTime = SystemClock.uptimeMillis() - startTime;
            //pausing
            if(loopTime<DELAY){
                try{
                    Thread.sleep(millis:DELAY-loopTime);
                }
                catch(InterruptedException e){
                    Log.e(tag:"Interrupted",msg:"Interrupted while sleeping");
                }
            }
        }
    }
}

File: app/src/main/java/com/example/myapplicationbird/GameThread.kt
package com.example.myapplicationbird

import android.graphics.Canvas
import android.os.SystemClock
import android.view.SurfaceHolder
import android.util.Log

class GameThread(private val surfaceHolder: SurfaceHolder) : Thread() {
    private var isRunning: Boolean = true
    private var startTime: Long = 0
    private var loopTime: Long = 0
    private val DELAY: Long = 33

    override fun run() {
        // Looping until the boolean is false
        while (isRunning) {
            startTime = SystemClock.uptimeMillis()
            val canvas = surfaceHolder.lockCanvas(null)
            if (canvas != null) {
                synchronized(surfaceHolder) {
                    AppConstants.getGameEngine().updateAndDrawBackgroundImage(canvas)
                    AppConstants.getGameEngine().updateAndDrawBird(canvas)
                    // Unlocking the canvas
                    surfaceHolder.unlockCanvasAndPost(canvas)
                }
            }
            // Loop time
            loopTime = SystemClock.uptimeMillis() - startTime
            // Pausing
            if (loopTime < DELAY) {
                try {
                    Thread.sleep(DELAY - loopTime)
                } catch (e: InterruptedException) {
                    Log.e("Interrupted", "Interrupted while sleeping")
                }
            }
        }
    }

    fun isRunning(): Boolean {
        return isRunning
    }

    fun setIsRunning(state: Boolean) {
        isRunning = state
    }
}